unittests/finality_proof.hpp

```diff
@@ -0,0 +1,289 @@
+#pragma once
+
+#include "finality_test_cluster.hpp"
+
+using mvo = mutable_variant_object;
+
+namespace finality_proof {
+
+   // data relevant to IBC
+   struct ibc_block_data_t {
+      signed_block_ptr block;
+      qc_data_t qc_data;
+      action_trace onblock_trace;
+      finality_data_t finality_data;
+      uint32_t active_finalizer_policy_generation;
+      uint32_t last_pending_finalizer_policy_generation;
+      uint32_t last_proposed_finalizer_policy_generation;
+      digest_type action_mroot;
+      digest_type base_digest;
+      digest_type active_finalizer_policy_digest;
+      digest_type last_pending_finalizer_policy_digest;
+      digest_type last_proposed_finalizer_policy_digest;
+      digest_type finality_digest;
+      digest_type computed_finality_digest;
+      digest_type afp_base_digest;
+      digest_type finality_leaf;
+      digest_type finality_root;
+   };
+
+   //generate a proof of inclusion for a node at index from a list of leaves
+   static std::vector<digest_type> generate_proof_of_inclusion(const std::vector<digest_type> leaves, const size_t index) {
+      auto _leaves = leaves;
+      auto _index = index;
+
+      std::vector<digest_type> merkle_branches;
+
+      while (_leaves.size()>1){
+         std::vector<digest_type> new_level;
+         for (size_t i = 0 ; i < _leaves.size() ; i+=2){
+            digest_type left = _leaves[i];
+
+            if (i + 1 < _leaves.size() && (i + 1 != _leaves.size() - 1 || _leaves.size() % 2 == 0)){
+               // Normal case: both children exist and are not at the end or are even
+               digest_type right = _leaves[i+1];
+
+               new_level.push_back(fc::sha256::hash(std::pair<digest_type, digest_type>(left, right)));
+               if (_index == i || _index == i + 1) {
+                 merkle_branches.push_back(_index == i ? right : left);

```

Comment:
Should it be `left : right`?

Reply:
When _index==i, we are processing a left node, so we have to provide its associated sibling (right node) as  the proof element. And vice-versa.